### PR TITLE
Sm/library-events

### PR DESCRIPTION
**Pull request by mshanemc**
part 1 of @W-9660828@

- modify lifecycle singleton to replace with newer version <-- please validate this logic!
- adds telemetry/warning events
- move process.emitWarning to the lifecycle.emitWarning in connection/sfdcUrl

### CODE REVIEW

File: src/lifecycleEvents.ts [hunk 1]
```diff
@@ -32,14 +36,20 @@ declare const global: {
...
+  public static readonly warningEventName = 'warning';
   private debug = Debug(`sfdx:${this.constructor.name}`);
-  private readonly listeners: Dictionary<callback[]>;
 
-  private constructor() {
-    this.listeners = {};
+  private constructor(private readonly listeners: Dictionary<callback[]> = {}) {}
+
+  public static staticVersion(): string {
+    return pjson.version;
```
Comment (shetzel) [inline src/lifecycleEvents.ts:52]: Should this be private?

File: src/lifecycleEvents.ts [hunk 2]
```diff
@@ -63,11 +73,28 @@ export class Lifecycle {
...
+      global.salesforceCoreLifecycle = new Lifecycle({ ...oldInstance.listeners });
+      // clean up any listeners on the old version
+      Object.keys(oldInstance.listeners).map((eventName) => {
+        oldInstance.removeAllListeners(eventName);
+      });
     }
     return global.salesforceCoreLifecycle;
   }
 
+  public version(): string {
```
Comment (shetzel) [inline src/lifecycleEvents.ts:94]: Should we js doc this?

**Review by shetzel — APPROVED** [review]
a few comments but nothing blocking

File: src/lifecycleEvents.ts (hunk 1 above)
Comment (mshanemc) [reply src/lifecycleEvents.ts:52]: I don't see the harm in public, and it makes the tests much simpler.

File: src/lifecycleEvents.ts (hunk 2 above)
Comment (mshanemc) [reply src/lifecycleEvents.ts:94]: yes.

Comment (mshanemc) [reply src/lifecycleEvents.ts:101]: yes, done.

File: src/lifecycleEvents.ts (hunk 1 above)
Comment (mshanemc) [reply src/lifecycleEvents.ts:55]: I'm not concerned about that, but I did use it in the tests.

**Pull request merged**